### DISCUSSION TITLE
How can we access the prompt given to the LLM?

### DISCUSSION THREAD

**Discussion by Chasearmer** [567-labs/instructor#350] [Q&A] [answered]
As the title suggests, how can we access the prompt that is generated from the Pydantic response model and passed to the LLM? I would be very interested in seeing what that looks, especially towards accessing the clarity of prompts created from nested Pydantic models.

**Comment by jxnl** [comment, marked as answer]
Making some improvements to docs

https://jxnl.github.io/instructor/concepts/logging/

**Reply by Chasearmer** [reply]
That's helpful to have, but I really meant the whole message (or list of messages) that are actually passed to the LLM. I noticed that the handle_response_model() function seems to add content beyond just what the model.model_json_schema() shows. I'm not quite tracking how it would all come together in the end just from visual inspection of this handle_response_model() function alone, and so having access to the new_kwargs['messages'] list would be really valuable. 

For example, it almost looks like the schema message is added twice in the final block of code
```
           # check that the first message is a system message
            # if it is not, add a system message to the beginning
            if new_kwargs["messages"][0]["role"] != "system":
                new_kwargs["messages"].insert(
                    0,
                    {
                        "role": "system",
                        "content": message,
                    },
                )

            # if the first message is a system append the schema to the end
            if new_kwargs["messages"][0]["role"] == "system":
                new_kwargs["messages"][0]["content"] += f"\n\n{message}"
```

Even in the case that I'm misreading the above, it would be helpful to see what it looks like when all the message formatting is done!

**Reply by jxnl** [reply]
if you set the logging level to debug openai should print everything!

**Reply by leobeeson** [reply]
@jxnl, I understand that the schema is passed to `new_kwargs["functions"]` or `new_kwargs["tools"]` depending on `Mode` (as per e.g. 
 [here](https://github.com/jxnl/instructor/blob/155b1a1441d12f0231bbe54e10e947cc388d744b/instructor/patch.py#L110)), and I can generate the schema using `response_model.model_json_schema()`. 

However, as I'm trying to emulate LangSmith locally, I've tried a couple of approaches for setting the logging level to debug in order to get the schema as passed to openai (e.g. logging `response`, `response._raw_response`, etc.), but was unable to get the schema logged this way. Could you please elaborate on "set the logging level to debug openai should print everything!"?

Thanks!

**Reply by Mr-Ruben** [reply]
> if you set the logging level to debug openai should print everything!

This is rather unclear/unspecific.

- Set it where? ( call / module )
- Of who? (instructor / openai )
- How?   (not everyone is familiar with python logging module)

---------------------

This got me on the right track
https://stackoverflow.com/questions/76256249/logging-in-the-open-ai-python-library



And after some experimentation (I have a nested logging setup)

Get my top logger  (this may be different for you)

```
toplog = logging.getLogger("<MyModuleName>")
toplog.handlers  # This should show something

```

This hooks openai logs to my own ones 

```

openai_loggers = [
    logger for logger in loggers if logger.name.startswith("openai")]

for openai_logger in openai_loggers:
    # Get the openai logger
    openai_logger = logging.getLogger(openai_logger.name)

    # Set the openai logger level to DEBUG
    openai_logger.setLevel(logging.DEBUG)

    # Add your logger's handlers to the openai logger
    for handler in toplog.handlers:
        openai_logger.addHandler(handler)
```


On the logs you will see the full body of the call



Extracting and formatting
```

'method': 'post', 
'url': '/chat/completions', 
'files': None, 
'json_data': {'messages': [{'role': 'user', 'content': 'Extract Jason is 25 years old'}], 
'model': 'gpt-3.5-turbo', 
'function_call': {'name': 'UserDetail'}, 
'functions': [
{'name': 'UserDetail', 'description': 'Correctly extracted `UserDetail` with all the required parameters with correct types', 
'parameters': {'properties': {'name': {'title': 'Name', 'type': 'string'}, 				
                                            'age': {'title': 'Age', 'type': 'integer'}}, 
'required': ['age', 'name'], 
'type': 'object'}}], 
'max_tokens': 1500, 'n': 1, 'temperature': 0.1}

```

https://i.imgur.com/GcYedZV.png









I hope this saves time to the next one.

**Reply by leobeeson** [reply]
Thank you, @Mr-Ruben! I managed to get a [working answer](https://discord.com/channels/1192334452110659664/1213176462941626458/1213204280480243762) on the Instructor Discord here, but your suggestion is even more straightforward (for those of us not familiar with python logging module). Thanks, again.

**Reply by jxnl** [reply]
Is anyone willing to help add this to docs? I'll make an issue

**Reply by leobeeson** [reply]
I'll take it! It'll be a great Python logging learning opportunity. Please assign the issue to me. Thanks.

**Reply by jxnl** [reply]
ty ty!

**Reply by pablospe** [reply]
Hi. Where the github issue can be found? Thanks!

**Reply by jss367** [reply]
The provided link returns a 404. For anyone else who finds this, my guess is the current version is here: https://python.useinstructor.com/concepts/logging/

**Reply by johnwlockwood** [reply]
[Logging docs](https://github.com/567-labs/instructor/blob/main/docs/concepts/logging.md)
